ratis-server/src/main/java/org/apache/ratis/server/JvmPauseMonitor.java

```diff
@@ -0,0 +1,181 @@
+                   + "pause of approximately " + extraSleepTime + "ms\n";
+    if (gcDiffs.isEmpty()) {
+      ret += "No GCs detected";
+    } else {
+      ret += Joiner.on("\n").join(gcDiffs);
+    }
+    return ret;
+  }
+
+  public JvmPauseMonitor(RaftServerProxy proxy) {
+    this.proxy = proxy;
+  }
+
+  private Map<String, GcTimes> getGcTimes() {
+    Map<String, GcTimes> map = Maps.newHashMap();
+    List<GarbageCollectorMXBean> gcBeans =
+      ManagementFactory.getGarbageCollectorMXBeans();
+    for (GarbageCollectorMXBean gcBean : gcBeans) {
+      map.put(gcBean.getName(), new GcTimes(gcBean));
+    }
+    return map;
+  }
+
+  private static class GcTimes {
+    private GcTimes(GarbageCollectorMXBean gcBean) {
+      gcCount = gcBean.getCollectionCount();
+      gcTimeMillis = gcBean.getCollectionTime();
+    }
+
+    private GcTimes(long count, long time) {
+      this.gcCount = count;
+      this.gcTimeMillis = time;
+    }
+
+    private GcTimes subtract(GcTimes other) {
+      return new GcTimes(this.gcCount - other.gcCount,
+        this.gcTimeMillis - other.gcTimeMillis);
+    }
+
+    @Override
+    public String toString() {
+      return "count=" + gcCount + " time=" + gcTimeMillis + "ms";
+    }
+
+    private long gcCount;
+    private long gcTimeMillis;
+  }
+
+  class Monitor implements Runnable {
+    private final String name = JvmPauseMonitor.this + "-" + JavaUtils
+                                                               .getClassSimpleName(getClass());
+
+    public void run() {
+      while (shouldRun) {
+        Stopwatch sw = Stopwatch.createUnstarted();
+        Map<String, GcTimes> gcTimesBeforeSleep = getGcTimes();
+        LOG.info("Starting Ratis JVM pause monitor");
+        while (shouldRun) {
+          sw.reset().start();
+          try {
+            Thread.sleep(SLEEP_INTERVAL_MS);
+          } catch (InterruptedException ie) {
+            return;
+          }
+          long extraSleepTime =
+            sw.elapsed(TimeUnit.MILLISECONDS) - SLEEP_INTERVAL_MS;
+          Map<String, GcTimes> gcTimesAfterSleep = getGcTimes();
+
+          if (extraSleepTime > 0) {
+            LOG.warn(formatMessage(extraSleepTime, gcTimesAfterSleep,
+              gcTimesBeforeSleep));
+          }
+          totalGcExtraSleepTime += extraSleepTime;
+          gcTimesBeforeSleep = gcTimesAfterSleep;
+          int leaderStepDownWaitTime =
+            RaftServerConfigKeys.LeaderElection.leaderStepDownWaitTime(
+              proxy.getProperties()).toIntExact(TimeUnit.MILLISECONDS);
+          int rpcSlownessTimeoutMs = RaftServerConfigKeys.Rpc.slownessTimeout(
+            proxy.getProperties()).toIntExact(TimeUnit.MILLISECONDS);
+          try {
+            if (totalGcExtraSleepTime > leaderStepDownWaitTime) {
+              proxy.getImpls().forEach(RaftServerImpl::stepDownOnJvmPause);
+            }
+            if (totalGcExtraSleepTime > rpcSlownessTimeoutMs) {
```

Comment:
They should use extraSleepTime instead of totalGcExtraSleepTime.

Reply:
If close proxy,  we do not need to step down. So
 ```
if (extraSleepTime > rpcSlownessTimeoutMs) {
   proxy.close();
 } else if (extraSleepTime > leaderStepDownWaitTime) {
  proxy.getImpls().forEach(RaftServerImpl::stepDownOnJvmPause);
 }
```